### PR TITLE
split up middleware, add OptionalPromise, add cache entry lifecycle

### PR DESCRIPTION
**Pull request by phryneas**
Background on the "cache entry lifecycle": https://github.com/rtk-incubator/rtk-query/issues/215

### CODE REVIEW

**Comment by phryneas** [comment]
what's still left todo here code-wise is removing the existing endpoint lifecycle onStart/etc.

File: src/query/core/buildMiddleware/queryLifecycle.ts [hunk 1]
```diff
@@ -40,11 +40,14 @@ export const build: SubMiddlewareBuilder = ({
           originalArgs
         )
 
+        const extra = mwApi.dispatch((_, __, extra) => extra)
```
Comment (markerikson) [inline src/query/core/buildMiddleware/queryLifecycle.ts:110]: 🤣

**Comment by phryneas** [comment]
Btw, Bikeshedding:

we need a discussion if `onCacheEntryAdded` and `onQuery` are appropriate names

File: src/query/core/buildMiddleware/queryLifecycle.ts (hunk 1 above)
Comment (phryneas) [reply src/query/core/buildMiddleware/queryLifecycle.ts:112]: Genius, isn't it? :D

I was like "the old API had it, so we mayybe need it here too?" and then "how on earth do I get that?". Yup.

**Comment by phryneas** [comment]
And internal TODO:

 I'm thinking of adding a `updateCacheEntry` (a pre-bound version of `updateQueryResult`) as an addition to https://github.com/reduxjs/redux-toolkit/pull/1034/commits/e2e1fbc0cdc461da92d4a4f936e3b04c393c0daf#diff-37992e5ad65f35286d92ad23730058800d8a872da260317ff5e9459aa03433a8R17-R41

Also, maybe adding a `undo` property on the thunk return value of `updateQueryResult` so that usually, `patchQueryResult` doesn't need to be dispatched by hand

File: src/query/core/buildMiddleware/cacheLifecycle.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+export const build: SubMiddlewareBuilder = ({
+  api,
+  reducerPath,
+  context,
+  queryThunk,
+  mutationThunk,
+}) => {
+  type CacheLifecycle = {
```
Comment (asherccohen) [inline src/query/core/buildMiddleware/cacheLifecycle.ts:120]: Can this be moved outside, or is there a reason it's declared here?

File: src/query/createApi.ts [hunk 3]
```diff
@@ -277,6 +277,26 @@ export function buildCreateApi<Modules extends [Module<any>, ...Module<any>[]]>(
             }
             x.providesTags ??= x.provides
           }
+          if (x.onStart || x.onSuccess || x.onError) {
```
Comment (asherccohen) [inline src/query/createApi.ts:280]: Nice, glad u added a message here!

File: src/query/core/buildMiddleware/cacheLifecycle.ts (hunk 2 above)
Comment (phryneas) [reply src/query/core/buildMiddleware/cacheLifecycle.ts:120]: It's only used in a locally scoped variable in the next line and once more to declare a value to add to that variable.

Moving it outside would raise questions like "should this be exported" etc, just keeping it here just circumvents all that and will also 100% remove it on transpilation.

File: src/query/tests/optimisticUpdates.test.tsx [hunk 4]
```diff
@@ -30,25 +30,19 @@ const api = createApi({
...
           api.util.updateQueryResult('post', id, (draft) => {
             Object.assign(draft, patch)
           })
-        ).inversePatches
-      },
-      onError({ id }, { dispatch, context }) {
-        dispatch(api.util.patchQueryResult('post', id, context.undoPost))
+        )
+        resultPromise.catch(undo)
       },
```
Comment (phryneas) [inline src/query/tests/optimisticUpdates.test.tsx:46]: It's way too late, but this diff is kinda worth it

File: src/query/core/buildMiddleware/cacheLifecycle.ts (hunk 2 above)
Comment (asherccohen) [reply src/query/core/buildMiddleware/cacheLifecycle.ts]: Cool. Tx for the explanation.

**Pull request merged**